File: pkg/integrations/launchdarkly/delete_flag.go
package launchdarkly

import (
	"fmt"
	"net/http"
	"strings"

	"github.com/google/uuid"
	"github.com/mitchellh/mapstructure"
	"github.com/superplanehq/superplane/pkg/configuration"
	"github.com/superplanehq/superplane/pkg/core"
)

const DeleteFlagPayloadType = "launchdarkly.deleteFlag.result"

type DeleteFlag struct{}

type DeleteFlagSpec struct {
	ProjectKey string `json:"projectKey" mapstructure:"projectKey"`
	FlagKey    string `json:"flagKey" mapstructure:"flagKey"`
}

type DeleteFlagOutput struct {
	Deleted    bool   `json:"deleted"`
	StatusCode int    `json:"statusCode"`
	Message    string `json:"message"`
}

func (d *DeleteFlag) Name() string {
	return "launchdarkly.deleteFlag"
}

func (d *DeleteFlag) Label() string {
	return "Delete Feature Flag"
}

func (d *DeleteFlag) Description() string {
	return "Permanently delete a LaunchDarkly feature flag"
}

func (d *DeleteFlag) Documentation() string {
	return `Deletes a LaunchDarkly feature flag from all environments.

## Warning
This operation is destructive and cannot be undone.

## Inputs
- Project Key
- Flag Key

## Output
Returns whether the flag was deleted, HTTP status code, and a status message.

A 404 response is handled as an expected "already deleted/not found" result.`
}

func (d *DeleteFlag) Icon() string {
	return "trash"
}

func (d *DeleteFlag) Color() string {
	return "#DC2626"
}

func (d *DeleteFlag) ExampleOutput() map[string]any {
	return deleteFlagExampleOutput()
}

func (d *DeleteFlag) OutputChannels(config any) []core.OutputChannel {
	return []core.OutputChannel{core.DefaultOutputChannel}
}

func (d *DeleteFlag) Configuration() []configuration.Field {
	return []configuration.Field{
		{
			Name:        "projectKey",
			Label:       "Project Key",
			Type:        configuration.FieldTypeString,
			Required:    true,
			Description: "Key of the LaunchDarkly project that owns the flag",
			Placeholder: "default",
		},
		{
			Name:        "flagKey",
			Label:       "Flag Key",
			Type:        configuration.FieldTypeString,
			Required:    true,
			Description: "Unique key of the LaunchDarkly feature flag",
			Placeholder: "my-feature-flag",
		},
	}
}

func (d *DeleteFlag) Setup(ctx core.SetupContext) error {
	spec := DeleteFlagSpec{}
	if err := mapstructure.Decode(ctx.Configuration, &spec); err != nil {
		return fmt.Errorf("failed to decode configuration: %w", err)
	}

	if strings.TrimSpace(spec.ProjectKey) == "" {
		return fmt.Errorf("projectKey is required")
	}
	if strings.TrimSpace(spec.FlagKey) == "" {
		return fmt.Errorf("flagKey is required")
	}

	return nil
}

func (d *DeleteFlag) ProcessQueueItem(ctx core.ProcessQueueContext) (*uuid.UUID, error) {
	return ctx.DefaultProcessing()
}

func (d *DeleteFlag) Execute(ctx core.ExecutionContext) error {
	spec := DeleteFlagSpec{}
	if err := mapstructure.Decode(ctx.Configuration, &spec); err != nil {
		return fmt.Errorf("failed to decode configuration: %w", err)
	}

	projectKey := strings.TrimSpace(spec.ProjectKey)
	flagKey := strings.TrimSpace(spec.FlagKey)

	if projectKey == "" {
		return fmt.Errorf("projectKey is required")
	}
	if flagKey == "" {
		return fmt.Errorf("flagKey is required")
	}

	client, err := NewClient(ctx.HTTP, ctx.Integration)
	if err != nil {
		return fmt.Errorf("failed to create LaunchDarkly client: %w", err)
	}

	statusCode, _, err := client.DeleteFlag(projectKey, flagKey)
	if err != nil {
		if statusCode == http.StatusNotFound {
			output := DeleteFlagOutput{
				Deleted:    false,
				StatusCode: http.StatusNotFound,
				Message:    fmt.Sprintf("Feature flag %q was not found in project %q.", flagKey, projectKey),
			}
			return ctx.ExecutionState.Emit(core.DefaultOutputChannel.Name, DeleteFlagPayloadType, []any{output})
		}
		return fmt.Errorf("failed to delete flag: %w", err)
	}

	output := DeleteFlagOutput{
		Deleted:    true,
		StatusCode: statusCode,
		Message:    fmt.Sprintf("Feature flag %q was successfully deleted from project %q.", flagKey, projectKey),
	}

	if statusCode != http.StatusNoContent && statusCode != http.StatusOK {
		output.Message = fmt.Sprintf("Delete request for feature flag %q returned status %d.", flagKey, statusCode)
	}

	return ctx.ExecutionState.Emit(core.DefaultOutputChannel.Name, DeleteFlagPayloadType, []any{output})
}

func (d *DeleteFlag) Actions() []core.Action {
	return []core.Action{}
}

func (d *DeleteFlag) HandleAction(ctx core.ActionContext) error {
	return nil
}

func (d *DeleteFlag) HandleWebhook(ctx core.WebhookRequestContext) (int, error) {
	return http.StatusOK, nil
}

func (d *DeleteFlag) Cancel(ctx core.ExecutionContext) error {
	return nil
}

func (d *DeleteFlag) Cleanup(ctx core.SetupContext) error {
	return nil
}

File: launchdarkly/actions/delete-flag.ts
import { createAction, ActionContext } from '@superplane/framework';
import {
  launchDarklyConnection,
  buildAuthHeaders,
  LD_API_BASE_URL,
  type LaunchDarklyConnectionProps,
} from '../auth';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * Input properties for the "Delete Feature Flag" action.
 */
export interface DeleteFlagProps {
  /**
   * The key of the LaunchDarkly project that owns the flag.
   *
   * @example "my-application"
   */
  projectKey: string;

  /**
   * The unique key of the feature flag to delete.
   *
   * @example "enable-dark-mode"
   */
  flagKey: string;
}

/**
 * Output of the "Delete Feature Flag" action.
 */
export interface DeleteFlagOutput {
  /** `true` when the flag was successfully deleted. */
  deleted: boolean;

  /** A human-readable status message. */
  message: string;

  /** HTTP status code returned by the API. */
  statusCode: number;
}

// ---------------------------------------------------------------------------
// Action definition
// ---------------------------------------------------------------------------

/**
 * **Delete Feature Flag** – REST API Action (DELETE)
 *
 * Permanently deletes a LaunchDarkly feature flag in **all environments**.
 *
 * > ⚠️ **Use with caution** – this action cannot be undone. Only delete
 * > flags that your application no longer references.
 *
 * A `404` response is handled gracefully, returning `deleted: false` with
 * a descriptive message rather than throwing.
 *
 * @see {@link https://launchdarkly.com/docs/api/feature-flags/delete-feature-flag | API – Delete Feature Flag}
 */
export const deleteFlag = createAction<
  DeleteFlagProps,
  DeleteFlagOutput,
  LaunchDarklyConnectionProps
>({
  name: 'Delete Feature Flag',
  description:
    'Permanently delete a LaunchDarkly feature flag from all environments. Use with caution.',
  connection: launchDarklyConnection,

  // ── Prop / input definitions ──────────────────────────────────────────
  props: {
    projectKey: {
      type: 'string',
      label: 'Project Key',
      description:
        'The key of the LaunchDarkly project that contains the flag.',
      required: true,
    },
    flagKey: {
      type: 'string',
      label: 'Flag Key',
      description:
        'The unique key of the feature flag to delete.',
      required: true,
    },
  },

  // ── Action handler ────────────────────────────────────────────────────
  /**
   * Execute the DELETE request against the LaunchDarkly API.
   *
   * @param ctx - Action context provided by the SuperPlane runtime,
   *              including the resolved connection and user-supplied props.
   * @returns A result indicating whether the flag was successfully deleted.
   */
  async run(
    ctx: ActionContext<DeleteFlagProps, LaunchDarklyConnectionProps>,
  ): Promise<DeleteFlagOutput> {
    const { projectKey, flagKey } = ctx.props;

    // Validate required inputs
    if (!projectKey?.trim()) {
      throw new Error(
        '[LaunchDarkly:deleteFlag] "projectKey" is required and cannot be empty.',
      );
    }
    if (!flagKey?.trim()) {
      throw new Error(
        '[LaunchDarkly:deleteFlag] "flagKey" is required and cannot be empty.',
      );
    }

    const url = `${LD_API_BASE_URL}/flags/${encodeURIComponent(projectKey)}/${encodeURIComponent(flagKey)}`;
    const headers = buildAuthHeaders(ctx.connection);

    console.debug(
      `[LaunchDarkly:deleteFlag] Deleting flag "${flagKey}" from project "${projectKey}"`,
    );

    try {
      const response = await fetch(url, {
        method: 'DELETE',
        headers,
      });

      // ── 204 – successful deletion (no content) ─────────────────────
      if (response.status === 204) {
        console.info(
          `[LaunchDarkly:deleteFlag] Flag "${flagKey}" successfully deleted (204).`,
        );
        return {
          deleted: true,
          message: `Feature flag "${flagKey}" was successfully deleted from project "${projectKey}".`,
          statusCode: 204,
        };
      }

      // ── 404 – flag does not exist ──────────────────────────────────
      if (response.status === 404) {
        console.info(
          `[LaunchDarkly:deleteFlag] Flag "${flagKey}" not found (404).`,
        );
        return {
          deleted: false,
          message: `Feature flag "${flagKey}" was not found in project "${projectKey}". It may have already been deleted.`,
          statusCode: 404,
        };
      }

      // ── 409 – conflict (e.g. flag is in use by an experiment) ──────
      if (response.status === 409) {
        const errorBody = await response.text();
        throw new Error(
          `Cannot delete flag "${flagKey}": conflict (409). ` +
            `The flag may be in use by an active experiment or approval request. ` +
            `Details: ${errorBody}`,
        );
      }

      // ── Other non-success statuses ─────────────────────────────────
      if (!response.ok) {
        const errorBody = await response.text();
        throw new Error(
          `LaunchDarkly API responded with status ${response.status}: ${errorBody}`,
        );
      }

      // ── Unexpected success status (e.g. 200) – still treat as OK ───
      return {
        deleted: true,
        message: `Feature flag "${flagKey}" deletion returned status ${response.status}.`,
        statusCode: response.status,
      };
    } catch (error: unknown) {
      // Re-throw errors we already created above
      if (
        error instanceof Error &&
        (error.message.startsWith('LaunchDarkly API') ||
          error.message.startsWith('Cannot delete flag'))
      ) {
        throw error;
      }

      const message =
        error instanceof Error ? error.message : String(error);
      throw new Error(
        `[LaunchDarkly:deleteFlag] Request failed: ${message}`,
      );
    }
  },
});
